JVC D-ILA Control Plugin

Civer666 opened the pull request:
Hey guys,
i want to add my [JVC D-ILA Control Plugin](https://knx-user-forum.de/forum/supportforen/smarthome-py/1188479-plugin-steuerung-von-jvc-d-ila-projektoren) to smarthomeNG plugins!
Is there a way to get write access, too?

Greetings, KNXinsHaus

ohinckel commented on jvcproj/README.md:
```diff
@@ -0,0 +1,251 @@
+
+## Notes
+I use the plugin to control my JVC DLA- X5000 and transfer raw gammatables generated with jvcprojectortools by arvehj. Many thanks to arvehjs groundwork so i could have a look in the communication procedure!
+You can use all operation commands given in the "JVC External Control Command Communication Specification". Wrong or unknown commands will be ignored by the plugin and the projector. Most of the commands won`t be acknowledged when the projector has no input signal. Request commands are not yet supported.
```
You should use simple quotes, not back ticks (could lead to problems when rendering this page).

ohinckel commented on jvcproj/README.md:
````diff
@@ -0,0 +1,251 @@
+```
+
+#### Attributes
+  * `host`: the host adress of the projector
````
Typo in "address", two times (see above).

ohinckel commented on jvcproj/README.md:
```diff
@@ -0,0 +1,251 @@
+        type: bool
+        visu_acl: rw
+        jvcproj_cmd: 21 89 01 00 00 0A
+        enforce_updates: 'True'
```
Do you really need these single quotes around "True"?

Civer666 replied:
No, i do not think so! I think these quotes exist in my items.conf since the beginning of smarthome.py! I will change it!

ohinckel commented on jvcproj/__init__.py:
```diff
@@ -0,0 +1,367 @@
+        """
+        if item():
+            if self.has_iattr(item.conf, 'jvcproj_cmd'):
+                if self.get_iattr_value(item.conf, 'jvcproj_cmd') == 'None':
```
Why do you need a special 'None'  check? Is the `has_iattr()` check one line above not enough? Do you really want to set the attribute to value `"None"`?

Civer666 replied:
smartplugin.py returns True if item exists. But when no command is declared (jvcproj_cmd = ..) the returned value will be "None" or not? So you can debug a message to have a look in the item because no command is given there. That my intension.

ohinckel commented on jvcproj/plugin.yaml:
```diff
@@ -0,0 +1,25 @@
+        de: 'Plugin um JVC D-ILA Projektoren zu kontrollieren und durch jvcprojectortools erstellte Gammakurven zu uebertragen'
+        en: 'plugin to control JVC D-ILA projectors and transfer gammatables generated with jvcprojectortools'
+    maintainer: NPortz
+    tester: 
```
Do you now any other testers to list here?

Civer666 replied:
No, sorry. Of course i know some People with that kind of projector. But most of them use Crestron to control the projector. I`m the only one who tested it at the moment.

ohinckel commented on jvcproj/plugin.yaml:
```diff
@@ -0,0 +1,25 @@
+#    support: https://knx-user-forum.de/forum/supportforen/smarthome-py
+
+    version: 1.0.0                 # Plugin version
+    sh_minversion: 1.1             # minimum shNG version to use this plugin
```
I guess SH 1.1 does not have SmartPlugin support. I would increase at least to 1.3.

Civer666 replied:
Yes, thats right! The complete plugin.yaml has to be revised anyway.

ohinckel commented on jvcproj/plugin.yaml:
```diff
@@ -0,0 +1,25 @@
+    classname: JVC_DILA_Control    # class containing the plugin
+
+parameters:
+    # Definition of parameters to be configured in etc/plugin.yaml
```
If you document the plugin parameters here you could benefit from some default values also declare here instead of checking in the code. Take a look into the [docs](http://smarthomeng.de/developer/development_plugin/plugin_metadata.html) to see how it works.

Civer666 replied:
Yes, i will do that!

ohinckel commented on jvcproj/plugin.yaml:
```diff
@@ -0,0 +1,25 @@
+    # Definition of parameters to be configured in etc/plugin.yaml
+    
+item_attributes:
+    # Definition of item attributes defined by this plugin
```
Adding documentation of the item attributes here would also help to improve the documentation.

Civer666 commented:
Ok. I´ve updated the Readme.md, added parameters to plugin.yaml and removed the option to set a port because it is always fixed to 20554 (since years) by JVC.
Only one thing: i wont set item_attributes to the plugin.yaml because the attributes are raw hex commands in variable lenght. It depends on beamer generation, too.

ohinckel commented on jvcproj/plugin.yaml:
```diff
@@ -19,7 +19,20 @@ plugin:
+            en: 'Define IP address of your JVC projector.'
+
+    gammaconf_dir:
+        type: dict
```
This should be type `str` because you configured a path, not a dictionary directly.

ohinckel commented:
Adding configuration it `item_attributes` is (currently) only used for documentation purpose. So I suggest to add them too. You don't need to provide a default value. Its enough to describe and declare the type of it in the `plugin.yaml`.

Civer666 commented:
Maybe a stupid question but i have to ask it! ;) Are "jvcproj_cmd" and "jvcproj_gamma" boolean or a string items? The item itself is boolean but the declared commands are strings! Normally i would say they're boolean...

ohinckel commented:
When you talk about "items" you mean smarthomeNG item or just item attributes? In the first case the `type` defines the type of the item's value and not the type of any item attribute.

```yaml
  items:
    cmd:
      type : bool
      jvcproj_cmd: XY XY XY
    gamma:
      type: bool
      jvcproj_gamma: ....
```

When you are talking abou the `item_attributes` section of `plugin.yaml` then you need to define the type of the attributes. In the case above both seems to be a `str`.

Civer666 commented:
Ok, in the end i did it right yesterday. I will push the changes this evening.

ohinckel commented:
I guess we can merge this now. Opinions?

Civer666 commented:
From my side: yes you can. The plugin works fine. I just haven`t found any other testers yet.
